### PR TITLE
Added doubleMem functionality for lsf jobs that die due to reaching imposed memory limit

### PR DESCRIPTION
**Pull request by drkennetz**
Often in genomics pipelines, especially for post-processing, memory limits of programs are not well-defined and can be dependent on data. This PR will add a memory doubling functionality if the flag `--doubleMem` is included on the command line for the LSF scheduler which will double the memory of and resubmit jobs that die due to reaching the LSF imposed memory limit. This functionality is incredibly useful for on-prem clusters. This addresses the issue opened by me: #3269

### CODE REVIEW

**Comment by adamnovak** [comment]
I don't think using a magic exit status and then rewriting the exit reason in the leader is the right way to do this. When the job runs out of memory, the LSF batch system should return an actually-correct `UpdatedBatchJobInfo` from `getUpdatedBatchJob` with the right exit reason set already.

I think what we want to do is change [the semantics defined for `getJobExitCode`](https://github.com/DataBiosphere/toil/blob/9a2cee472d0e2da2fb391314b1d1a9d666ad1425/src/toil/batchSystems/abstractGridEngineBatchSystem.py#L281), so that it can return both the code and the reason (or maybe so that it can return either the code or an instance of `BatchJobExitReason`, which I think we could tell apart with `isinstance`), and then update [here](https://github.com/DataBiosphere/toil/blob/9a2cee472d0e2da2fb391314b1d1a9d666ad1425/src/toil/batchSystems/abstractGridEngineBatchSystem.py#L197) to fill in the reason if it is available, and (if adding more return values) fix the other `getJobExitCode` implementations for the other grid engine batch systems to still work.

File: src/toil/common.py
Comment (adamnovak) [inline src/toil/common.py:559]: Since this adds another option it should also update https://github.com/DataBiosphere/toil/blob/c526b6fc20a81b9e04bad46d14840511666d8c3e/docs/running/cliOptions.rst

File: src/toil/jobGraph.py
Comment (adamnovak) [inline src/toil/jobGraph.py:155]: It looks like memory can be both doubled and upped to the default in the same call. That might confuse people; it should do only one or the other.

**Review by adamnovak — CHANGES_REQUESTED** [review]
In addition to changing how the reason is arrived at, I found a couple other things I think we want to change before taking this.

File: src/toil/jobGraph.py
Comment (drkennetz) [reply src/toil/jobGraph.py:155]: This is true, if the default is greater than double the step's memory, it will be bumped to the default's memory. I actually thought that behavior seemed okay. What will end up happening if the memory is like 500M and the default is 2G and a job fails due to memory:

1a. 500M -> 1G from the memory double
1b. 1G -> 2G due to default
 Job Fails again due to memory:
2a. 2G -> 4G
Job Fails again:
3a. 4G -> 8G
...
I can make these exclusive conditions if you think that would be better.

**Comment by drkennetz** [comment]
@adamnovak yeah I didn't really think this was the ideal way to go about this, but I think I was still missing some pieces of the puzzle when trying to implement. I figured with the PR you might be able to give me some valuable feedback to finish the changes, and I was right! Thanks for the review. I'll go ahead and make the recommended changes and request a review.

**Comment by drkennetz** [comment]
@adamnovak I went with the logic of returning an instance of `BatchJobExitReason` in lsf then adding some additional logic rather than increasing the number of return values. This gets rid of the magic exit code and doesn't really change the flow. I added:
```
if exitStatus == 0 and exitReason == None:
```
because the exitReason is being captured now. This isn't entirely necessary because I assign an exitCode of 1 if the return type isinstance BatchJobExitReason simply because I am no longer returning the exit code by lsf and I did want to maintain the fact that there was actually an exit code. Let me know what you think.

**Review by adamnovak — CHANGES_REQUESTED** [review]
This looks great! I think it still needs a little documentation here on the new getJobExitCode semantics, though:

https://github.com/DataBiosphere/toil/blob/37d5e395c2a518d86f4b91b274cd693774d6ae09/src/toil/batchSystems/abstractGridEngineBatchSystem.py#L285-L291

We should say `:rtype: int|toil.batchSystems.abstractBatchSystem.BatchJobExitReason`, and mention in the text that it gets either the exit code, or a BatchJobExitReason if something else happened to the job other than exiting. Since that new behavior isn't really suggested by the function name, it's important to document it.

**Comment by adamnovak** [comment]
I've pulled this in for CI testing.

**Comment by drkennetz** [comment]
@adamnovak That's awesome! Hoping for the best. Should I wait for the CI to finish before updating the docs? I think you make a good point. Not sure how pushing while CI is running will affect things.

**Comment by drkennetz** [comment]
@adamnovak, I have updated the docs for `getJobExitCode`.

**Review by adamnovak — APPROVED** [review]
This all looks good to me!

**Review by DailyDreaming — APPROVED** [review]
LGTM.  Thanks for adding this!

**Comment by drkennetz** [comment]
@DailyDreaming no problem! I think it could be a useful feature for all GridEngines and relatively easy to implement for other schedulers if they report failure due to imposed mem limits.

Do you guys resolve merge conflicts? I'm seeing that we have one now. Unlikely that they touch the same part of the file but maybe.

**Comment by mr-c** [comment]
Worse case one of the committers can resolve the conflict, but if you don't mind doing it yourself, it would be appreciated!

**Comment by mr-c** [comment]
I'm guessing https://github.com/DataBiosphere/toil/pull/3250 was the culprit :-P

**Comment by drkennetz** [comment]
Since this PR is lsf specific and I know I can test it, I'll resolve it myself. It might take a second because I have to understand what was changed by converging these classes... haha. It shouldn't be too big of a problem though.

**Comment by drkennetz** [comment]
Would it be easier for us all if I just closed this PR and opened a new one?

**Pull request closed without merging**
